subsys/net/lib/dns/mdns_responder.c

```diff
@@ -606,6 +616,268 @@ static int dns_read(int sock,
+		NET_DBG("Cannot initialize semaphore (%d)", ret);
+		errno = -ret;
+		ret = DNS_EAI_SYSTEM;
+		goto out;
+	}
+
+	NET_DBG("Sending mDNS probe for %s", query);
+
+	/* Then the actual probe sending. RFC 6762 ch 8.1 says to use
+	 * the ANY resource type. Use the internal resolve function that does
+	 * not use cache.
+	 */
+	ret = dns_resolve_name_internal(&ctx->probe_ctx, query, DNS_RR_TYPE_ANY,
+					&dns_id, probe_cb, &ud, PROBE_TIMEOUT,
+					false);
+	if (ret < 0) {
+		NET_DBG("Cannot send mDNS probe (%d)", ret);
+		errno = -ret;
+		ret = DNS_EAI_SYSTEM;
+		goto fail;
+	}
+
+	/* Make sure that the semaphore is released even if the probe_cb
+	 * is not called and that it is longer that the probe timeout.
+	 */
+	ret = k_sem_take(&ud.wait_data, K_MSEC(PROBE_TIMEOUT + 50));
+	if (ret < 0) {
+		NET_DBG("mDNS probe timeout (%d)", ret);
+		errno = -ret;
+		ret = DNS_EAI_SYSTEM;
+
+		ret = dns_resolve_cancel_with_name(&ctx->probe_ctx, dns_id, query,
+						   DNS_RR_TYPE_ANY);
+		if (ret < 0) {
+			NET_DBG("Resolve cancel failed (%d)", ret);
+			errno = -ret;
+			ret = DNS_EAI_SYSTEM;
+		}
+	}
+
+	if (ud.status == DNS_EAI_ALLDONE) {
+		NET_WARN("mDNS probe received data, will not use \"%s\" name.", query);
+
+		failed_probes++;
+		errno = EEXIST;
+		ret = DNS_EAI_SYSTEM;
+
+		/* If some host had the same name as what we want to use,
+		 * then we will stop using this name.
+		 */
+		goto fail;
+	}
+
+	ret = 0;
+
+fail:
+	dns_resolve_close(&ctx->probe_ctx);
+out:
+	return ret;
+}
+
+static void cancel_probes(struct mdns_responder_context *ctx)
+{
+#if defined(CONFIG_NET_IPV4)
+	ARRAY_FOR_EACH(v4_ctx, i) {
+		if (&v4_ctx[i] == ctx) {
+			continue;
+		}
+
+		(void)k_work_cancel_delayable(&v4_ctx[i].probe_timer);
+	}
+#endif /* defined(CONFIG_NET_IPV4) */
+
+#if defined(CONFIG_NET_IPV6)
+	ARRAY_FOR_EACH(v6_ctx, i) {
+		if (&v6_ctx[i] != ctx) {
+			continue;
+		}
+
+		(void)k_work_cancel_delayable(&v6_ctx[i].probe_timer);
+	}
+#endif /* defined(CONFIG_NET_IPV6) */
+}
+
+static void probing(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct mdns_responder_context *ctx =
+		CONTAINER_OF(dwork, struct mdns_responder_context, probe_timer);
+	int ret;
+
+	if (failed_probes > 0) {
+		/* No point doing any probe because one already failed */
+		NET_DBG("Probe already failed, not sending any more probes.");
+
+		(void)k_work_cancel_delayable(&init_listener_timer);
+		return;
+	}
+
+	ret = send_probe(ctx);
+	if (ret < 0) {
+		if (errno == EEXIST) {
+			/* Cancel the other possible probes */
+			cancel_probes(ctx);
+		}
+
+		NET_DBG("Cannot send %s mDNS probe (%d, errno %d)",
+			ctx->dispatcher.local_addr.sa_family == AF_INET ? "IPv4" :
+			(ctx->dispatcher.local_addr.sa_family == AF_INET6 ? "IPv6" :
+			 ""),
+			ret, errno);
+	}
+}
+
+static void mdns_conn_event_handler(struct net_mgmt_event_callback *cb,
+				    uint32_t mgmt_event, struct net_if *iface)
+
+{
+	if (mgmt_event == NET_EVENT_L4_CONNECTED) {
```

Comment:
I'm not sure if connection manager events are a good choice here, note that the events are not tightly bound to a specific interface, I. e. if there are for example two interfaces in the system, one remains up and running, and the other one toggles up and down, there will be no event triggered. And if I understand correctly, the second interface should send a probe whenever it changes its state. Perhaps sticking to NET_IF_UP/DOWN events would be a better choice here?

Reply:
It is just that if up/down is probably not good here because the network interface needs to have IP address before the probing can be done. Connection manager has these IPv6/IPv4 connected events, perhaps I could utilize those here.

Reply:
After some initial testing, the `L4_IPVx_CONNECTED` is not good either as it is only triggered once. I will need to figure out something else for this, probably need to listen the IP address add signals and work from there.

Reply:
Maybe monitor NET_IF_UP/DOWN **and** IP_ADDR_ADD/REMOVE events, but check `net_if_ipv4/6_get_global_addr()` and send probe only if it gives non-NULL result? Too bad we've dropped the idea of interface-specific connectivity events some time ago.